gems/smithy-client/lib/smithy-client/plugins/protocols.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Client
+    module Plugins
+      # @api private
+      class Protocols < Plugin
+        option(
+          :protocol,
+          doc_default: '<DEFAULT_PROTOCOL>',
+          doc_type: 'String, Class',
+          docstring: 'The protocol to use for request serialization and response deserialization.'
+        )
+
+        def add_handlers(handlers, config)
+          return unless config.protocol
+
+          handlers.add(BuildHandler)
+          handlers.add(ParseHandler, step: :parse)
+        end
+
+        def before_initialize(client_class, options)
+          protocol = options[:protocol]
+
+          case protocol
+          when nil
+            protocol_class = client_class.protocols.values.first
+            options[:protocol] = protocol_class.new if protocol_class
+          when String
+            protocol_class = client_class.protocols[protocol]
+            raise ArgumentError, "Unknown protocol: #{protocol}" unless protocol_class
```

Comment:
Good call on adding this - we want to only support what protocols are supported by the model/server. I like that the Client file will contain these bring these details over.